0000-macro-value.md

```diff
@@ -0,0 +1,300 @@
+- Feature Name: `macro_value`
+- Start Date: 2021-03-27
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a new macro_rules matcher, `$name:value`, with identical semantics to that of a function capture.
+
+# Motivation
+[motivation]: #motivation
+
+Value arguments to function-like macros are tricky to deal with.
+While macro_rules macros don't suffer from the common and most egregious pitfalls of C-style preprocessor macros,
+such as misnested brackets and operator precedence,
+using an `$:expr` capture more than once still evaluates the expression more than once,
+duplicating side effects.
+
+Additionally,
+we have the additional wrinkle of the lifetime and drop timing of temporaries complicating matters further,
+if your intent is to write a macro invocation with equivalent-to-function-call semantics.
+Suffice to say,
+`let arg = $arg;` has the incorrect drop behavior,
+and the current best practice is to instead expand to
+
+```rust
+match ( $arg0, $arg1, ) {
+    ( arg0, arg1, ) => { /* macro body */ }
+}
+```
+
+instead.
+We can simplify this and make getting the correct behavior easier on macro authors.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+(in a section explaining `macro_rules` matchers:)
+
+While `$:expr` is good for capturing an expression and copying that expression into the macro-expanded code,
+it does exactly that: it duplicates the captured expression to every expansion of the capture.
+If, for example, you wrote the trivial `min!` macro,
+
+```rust
+macro_rules! min {
+    ( $a:expr, $b:expr ) => {
+        if $a <= $b { $a } else { $b }
+    };
+}
+```
+
+then both `$a` and `$b` are evaluated twice,
+once at each expansion point,
+as opposed to a single time,
+as would be the case if [`min` were a function](https://doc.rust-lang.org/std/cmp/fn.min.html).
+If you want the arguments to the macro to be evaluated a single time,
+as if they were simple function arguments,
+you can use the `$:value` matcher:
+
+```rust
+macro_rules! min {
+    ( $a:value, $b:value ) => {
+        if $a <= $b { $a } else { $b }
+    };
+}
+```
+
+This time,
+`$a` and `$b` are evaluated a single time upon invoking the macro,
+and each expansion of the capture refers to the same value,
+just like function arguments.
+
+(no need to mention temporary lifetimes in the guide.)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A new macro matching mode,
+`$:value`,
+is added.
```

Comment:
Not sure why there are so many new lines across the source. It makes it hard to read in non-rendered mode (which partially defeats the purpose of Markdown).

Reply:
A pattern I've seen is to use newlines after punctuation. The reasoning there is to reduce the diff noise of subsequent changes, as English prose often changes inside a clause, but less frequently between it. Using "arbitrary" newlines (e.g. at 80 columns) leads to a cascading series of diffs if you add enough words to an early line.

Reply:
A smarter diff for Markdown in GitHub would be useful :)

Reply:
Yeah, and actually the rich diff does pretty okay at it. I probably did too much newlines in this document, but there's a happy medium somewhere. In any case, completely unwrapped text is worse than overly wrapped (in my opinion), and maintaining "reasonable" wrapping in a consistently updated document is annoying.

It also helps me pay attention to my clause length. If I'm just writing without thinking about it, I tend to get super long sentences, with (potentially many) levels of indirection, and it becomes difficult to read -- so I have ended up with some (kludgy but workable) workflow things to help balance it.<sup>[1]</sup>

[1]: yes this mess of a run-on was on purpose to illustrate the point